lightning-invoice/src/payment.rs

```diff
@@ -188,10 +189,80 @@ where
 	logger: L,
 	event_handler: E,
 	/// Caches the overall attempts at making a payment, which is updated prior to retrying.
-	payment_cache: Mutex<HashMap<PaymentHash, PaymentAttempts<T>>>,
+	payment_cache: Mutex<HashMap<PaymentHash, PaymentInfo<T>>>,
 	retry: Retry,
 }
 
+/// Used by [`InvoicePayerUsingTime::payment_cache`] to track the payments that are either
+/// currently being made, or have outstanding paths that need retrying.
+struct PaymentInfo<T: Time> {
+	attempts: PaymentAttempts<T>,
+	paths: Vec<Vec<PaymentInfoRouteHop>>,
+}
+
+/// Subset of properties that is represented in `RouteHop`, so we only store what we need in
+/// [`InvoicePayerUsingTime::payment_cache`]
+///
+/// [`RouteHop`]: crate::routing::router::RouteHop
+#[derive(Clone, PartialEq, Eq)]
+struct PaymentInfoRouteHop {
+	/// The node_id of the node at this hop.
+	pub pubkey: PublicKey,
+	/// The channel that should be used from the previous hop to reach this node.
+	pub short_channel_id: u64,
+	/// The fee taken on this hop (for paying for the use of the *next* channel in the path).
+	/// For the last hop, this should be the full value of the payment (might be more than
+	/// requested if we had to match htlc_minimum_msat).
+	pub fee_msat: u64,
+}
+
+impl<T: Time> PaymentInfo<T> {
+	fn new() -> Self {
+		PaymentInfo {
+			attempts: PaymentAttempts::new(),
+			paths: vec![],
+		}
+	}
+}
+
+/// Used to store information about all the HTLCs that are inflight across all payment attempts
+struct AccountForInFlightHtlcs<'a, S: Score> {
+	scorer: &'a mut S,
+	/// Maps a channel's short channel id and its direction to the liquidity used up.
+	inflight_htlcs: HashMap<(u64, bool), u64>,
+}
+
+impl<'a, S: Score> Score for AccountForInFlightHtlcs<'a, S> {
+	fn channel_penalty_msat(&self, short_channel_id: u64, source: &NodeId, target: &NodeId, usage: ChannelUsage) -> u64 {
+		if let Some(used_liqudity) = self.inflight_htlcs.get(&(short_channel_id, source > target)) {
+			let usage = ChannelUsage {
+				inflight_htlc_msat: usage.inflight_htlc_msat + used_liqudity,
```

Comment:
Could you add test coverage for this? No test fails if I comment out this line. You'll likely need to do something fancy like make `TestRouter::find_route` call the the scorer and check some expectations that you set on `TestRouter` prior to using it.